added awsRequest for aws-sdk call backoff and retry in customResources

pratik-vii opened the pull request:
<!-- ⚠️⚠️ Acknowledge ALL below remarks -->
<!-- ⚠️⚠️ PR will not be processed if it doesn't meet outlined criteria -->

<!-- ⚠️⚠️ Do not propose PR's without prior agreement on solution in corresponding issue -->
<!-- ⚠️⚠️ Only documentation updates and obvious bug fixes are welcome without it -->

<!--
⚠️⚠️ Ensure to follow code style guidelines
https://github.com/serverless/serverless/blob/master/CONTRIBUTING.md#code-style
-->

<!--
⚠️⚠️ Ensure to cover changes with tests written according to test guidelines
https://github.com/serverless/serverless/blob/master/test/README.md
-->

<!-- ⚠️⚠️ Ensure that support for Node.js v6 is maintained. -->

<!--
⚠️⚠️ Ensure that proposed change passes CI. Confirm on that by running following scripts:
• npm run prettier-check
• npm run lint
• npm test
-->

<!--
⚠️⚠️ If proposed change touches integration with AWS services, confirm integration tests pass:
https://github.com/serverless/serverless/blob/master/test/README.md#aws-integration-tests
-->

<!-- ⚠️⚠️ After your PR is submitted, review the final CI status and address eventual failure -->

<!-- ⚠️⚠️ Answer below questions -->

<!--
Q1: Provide link to corresponding issue

• If PR *partially* addresses issue, ensure to rename "Closes" to "Addresses" ("Closes" term will automatically close an issue on PR merge)
• If it's a documentation update or obvious bug fix that has no corresponding issue, replace this line with short description of made changes
-->

Addresses: #8333

medikoo reviewed (changes requested):
@pratik-vii thanks for giving it a spin.

In given form this change is not effective. Note, that no custom resource logic was configured to new utility.  We need to ensure that all of the functions rely on this utility, instead of invoking `aws-sdk` directly.

Also this implementation does not look complete. I believe `serviceInstance` needs to be resolved on spot, as it's here: https://github.com/serverless/test/blob/7dfc92e1d0d93c91f1c9affc872afa41fcb1c89b/aws-request.js

pratik-vii commented:
>In given form this change is not effective
I was planning to integrate it once signature is approved, I had doubt to whether to keep `aws-sdk` instantiation in resource or utility itself. Which you answered already!
> I believe `serviceInstance` needs to be resolved on spot, as it's here:
Sure. Will do. I was thinking of doing the same pattern, but I thought to give responsibility of configuring the `aws-sdk` instance to resource.

medikoo commented on lib/plugins/aws/customResources/resources/apiGatewayCloudWatchRole/handler.js:
```diff
@@ -19,8 +18,9 @@ async function create(event, context) {
-  const apiGateway = new ApiGateway({ region });
-  const assignedRoleArn = (await apiGateway.getAccount().promise()).cloudwatchRoleArn;
+  const assignedRoleArn = (
+    await awsRequest({ name: 'APIGateway', params: { region } }, 'getAccount').promise()
```
`awsRequest` should return promise directly, there should be no need for `promise()`

medikoo commented on lib/plugins/aws/customResources/resources/utils.js:
```diff
@@ -90,11 +92,53 @@ function wait(ms) {
+  }
+);
+
+function awsRequest(service, method, ...args) {
```
Let's use async/await

medikoo commented on lib/plugins/aws/customResources/resources/utils.js:
```diff
@@ -1,5 +1,7 @@
 'use strict';
 
+const AWS = require('aws-sdk');
+const _ = require('lodash');
```
We should not introduce any dependencies here (note this is packaged for lambda, and we want to avoid `npm install`)

pratik-vii replied:
Yeah. Will remove the usage of `lodash`.

medikoo reviewed (changes requested):
Thank you @pratik-vii  for update. It looks way better now.

Please see my comments. Also please test it by running [integration tests](https://github.com/serverless/serverless/tree/master/test#aws-integration-tests), it's the only way to confirm, we did not  break anything with this refactor

pratik-vii commented:
@medikoo Thanks for the feedback.

I have updated the PR,
* Removed usage of `lodash`
* Changed `awsRequest` function to `async` function

I have tested the changes with dependent integration testcases
apiGateway.test.js
cognitoUserPool.test.js
eventBridge.test.js
s3.test.js

medikoo commented on lib/plugins/aws/customResources/resources/apiGatewayCloudWatchRole/handler.js:
```diff
@@ -68,40 +64,26 @@ async function create(event, context) {
-        // temporary state where just created AWS role is not being ready for use (yet)
-        await wait(10000);
-        return updateAccount(++counter);
-      }
```
I think we need to keep that try/catch, as in this case we want to catch also non-retryable errors, while `awsRequest` retries only if error is retryable

medikoo commented on lib/plugins/aws/customResources/resources/utils.js:
```diff
@@ -90,11 +91,53 @@ function wait(ms) {
+            requestTry < MAX_AWS_REQUEST_TRY &&
+            error &&
+            (error.statusCode === 429 ||
+              (error.retryable && error.statusCode !== 403 && error.code !== 'CredentialsError'))
```
I think we can limit this to `(error.statusCode === 429 || error.retryable)`, as in Lambda we should not expect credential errors, so it's not the case to consider

medikoo commented on lib/plugins/aws/customResources/resources/utils.js:
```diff
@@ -90,11 +91,53 @@ function wait(ms) {
+            (error.statusCode === 429 ||
+              (error.retryable && error.statusCode !== 403 && error.code !== 'CredentialsError'))
+          ) {
+            return wait(4000 + Math.random() * 3000).then(() => callAws(++requestTry));
```
7 seconds, is a lot in lambda, let's make it maybe `2000 + Math.random() * 1000`, and maybe increase `DEFAULT_MAX_AWS_REQUEST_TRY` to `8`

medikoo commented on lib/plugins/aws/customResources/resources/utils.js:
```diff
@@ -90,11 +91,53 @@ function wait(ms) {
+  const callAws = requestTry => {
+    return serviceInstance[method](...args)
+      .promise()
+      .then(
```
Let's use async/await

medikoo reviewed (changes requested):
Thank you @pratik-vii , that looks very good! I have just few minor suggestions. Let me know what you think

medikoo reviewed (approved):
Thank youi @pratik-vii ! That looks great
